src/lib/tzcode/gen-zone-abbrevs.pl

```diff
@@ -0,0 +1,207 @@
+#!/usr/bin/perl
+use strict;
+use warnings;
```

Comment:
In the commit message you say:
>    We use c-dt derived script to generate file describing timezone and their fixed ids.

However, commit contains not original versions of `zone.pl` and `zone.txt` ([source](https://github.com/tarantool/c-dt/tree/tarantool-master-v1/tools)). There is a difference for original versions and commited versions. Please describe added changes in commit message.

```diff
--- zone.txt	2022-04-23 08:10:47.624546506 +0300
+++ src/lib/tzcode/zone-abbrevs.txt	2022-04-23 08:02:28.222073173 +0300
@@ -72,9 +72,15 @@
 CET    +01:00          # Central European Time (Europe / Africa / Antarctica)
 CHADT  +13:45          # Chatham Island Daylight Time (Pacific)
 CHAST  +12:45          # Chatham Island Standard Time (Pacific)
+CHODST +09:00          # Choibalsan Daylight Saving Time (Asia)
+CHODT  +09:00          # Choibalsan Daylight Time (Asia)
+CHOST  +09:00          # Choibalsan Summer Time (Asia)
 CHOT   +08:00          # Choibalsan Time (Asia)
 ChST   +10:00          # Chamorro Standard Time (Pacific)
 CHUT   +10:00          # Chuuk Time (Pacific)
+CIDST  -04:00          # Cayman Islands Daylight Saving Time (Caribbean)
+CIST   -05:00          # Cayman Islands Standard Time (Caribbean)
+CIT    -05:00          # Cayman Islands Time (Caribbean)
 CKT    -10:00          # Cook Island Time (Pacific)
 CLDT   -03:00          # Chile Daylight Time (South America / Antarctica)
 CLST   -03:00          # Chile Summer Time (South America / Antarctica)
@@ -156,6 +162,9 @@
 HNP    -08:00          # Heure Normale du Pacifique (French) (North America)
 HNR    -07:00          # Heure Normale des Rocheuses (French) (North America)
 HNT    -03:30          # Heure Normale de Terre-Neuve (French) (North America)
+HOVDST +08:00          # Hovd Daylight Saving Time (Asia)
+HOVDT  +08:00          # Hovd Daylight Time (Asia)
+HOVST  +08:00          # Hovd Summer Time (Asia)
 HOVT   +07:00          # Hovd Time (Asia)
 HST    -10:00          # Hawaii Standard Time (North America / Pacific)
 I      +09:00          # India Time Zone (Military)
@@ -219,7 +228,8 @@
 NAPST  -08:00          # North American Pacific Standard Time (North America)
 NCT    +11:00          # New Caledonia Time (Pacific)
 NDT    -02:30          # Newfoundland Daylight Time (North America)
-NFT    +11:30          # Norfolk Time (Australia)
+NFT    +11:00          # Norfolk Time (Australia)
+NFDT   +12:00          # Norfolk Island Daylight Time Australia
 NOVST  +07:00          # Novosibirsk Summer Time (Asia)
 NOVT   +06:00          # Novosibirsk Time (Asia)
 NPT    +05:45          # Nepal Time (Asia)
@@ -254,7 +264,9 @@
 PT     -08:00          # Pacific Time (North America)
 PWT    +09:00          # Palau Time (Pacific)
 PYST   -03:00          # Paraguay Summer Time (South America)
+PYST   +08:30          # Pyongyang Standard Time (Asia)
 PYT    -04:00          # Paraguay Time (South America)
+PYT    +08:30          # Pyongyang Time (Asia)
 Q      -04:00          # Quebec Time Zone (Military)
 QYZT   +06:00          # Qyzylorda Time (Asia)
 R      -05:00          # Romeo Time Zone (Military)
@@ -281,9 +293,12 @@
 TKT    +13:00          # Tokelau Time (Pacific)
 TLT    +09:00          # East Timor Time (Asia)
 TMT    +05:00          # Turkmenistan Time (Asia)
+TOST   +14:00          # Tonga Summer Time (Pacific)
 TOT    +13:00          # Tonga Time (Pacific)
+TRT    +03:00          # Turkey Time (Asia)
 TVT    +12:00          # Tuvalu Time (Pacific)
 U      -08:00          # Uniform Time Zone (Military)
+ULAST  +09:00          # Ulan Bator Summer Time (Asia)
 ULAT   +08:00          # Ulaanbaatar Time (Asia)
 UT     +00:00          # Universal Time
 UTC    +00:00          # Coordinated Universal Time (Worldwide)
```

```diff
--- zone.pl	2022-04-23 08:10:38.796502463 +0300
+++ ./src/lib/tzcode/gen-zone-abbrevs.pl	2022-04-23 08:02:28.218073153 +0300
@@ -2,6 +2,25 @@
 use strict;
 use warnings;
 
+my $next_id    = 0;
+my $last_first = undef;
+
+# simply increment id for each next string passed
+# but jump to the nearest modulo 16 number if there
+# was first character changed - to make possible
+# manual inserts in future.
+sub next_id {
+    my ($string) = @_;
+    my $first = substr( $string, 0, 1 );
+    $next_id++;
+    return $next_id if length($string) == 1;    # sequential military zones
+    return $next_id if defined $last_first and $first eq $last_first;
+    $last_first = $first;
+    $next_id += 7;
+    $next_id &= ~0x7;   # jump to the next modulo of 8 for the next alpha letter
+    return $next_id;
+}
+
 sub encode {
     my ($string) = @_;
     my $v = 0;
@@ -41,6 +60,20 @@
     PST => -8*60, PDT => -7*60,
 );
 
+my %Regions = (
+    Africa      => 'f',
+    America     => 'm',
+    Antarctica  => 'n',
+    Asia        => 's',
+    Atlantic    => 't',
+    Australia   => 'u',
+    Etc         => 'e',
+    Europe      => 'r',
+    Factory     => 'c',
+    Indian      => 'i',
+    Pacific     => 'p',
+);
+
 my %FlagName;
 
 BEGIN {
@@ -54,73 +87,121 @@
     %FlagName = reverse %flags;
 }
 
+my $zone_abbrevs_file = shift || 'zone-abbrevs.txt';
+my $tzdata_zi_file    = shift || 'main.zi';
+
+# ACT    -05:00          # Acre Time (South America)
+# ACT    +10:30 / +09:30 # Australian Central Time (Australia)
+# ACWST  +08:45          # Australian Central Western Standard Time (Australia)
+
 my $EntryRx = do {
+    my $name   = '[A-Z][A-Za-z]{0,5}';
     my $offset = '[+-] (?: [01][0-9] | [2][0-3]) [:] [0-5][0-9]';
-    my $name   = '[A-Z][A-Za-z]{0,4}';
     qr< ($name) \s+ ($offset) (?: \s+ / \s+ ($offset))? \s+ #>xo
 };
 
-open(my $fh, '<:raw', 'tools/zone.txt')
-  or die qq<Could not open 'tools/zone.txt' for reading: '$!'>;
+my %ZoneAbbrevs;
 
-my %Zone;
+# read zone-abbrevs.txt with definition of all currently known
+# timezone abbreviations.
+sub read_abbrevs_file($) {
+    my $filename = shift;
+    open( my $fh, '<:raw', $filename )
+      or die qq<Could not open '$filename' for reading: '$!'>;
+
+    while (<$fh>) {
+        next if /\A \s* \# /x;
+
+        my ( $name, $offset1, $offset2 ) = m< \A $EntryRx >x
+          or die qq/Could not parse zone entry at line $_: $./;
+
+        my $encoded = encode($name);
+        my $offset  = 0;
+        my $flags   = 0;
 
-while (<$fh>) {
-    next if /\A \s* \# /x;
+        if ($offset2) {
+            $flags = DT_ZONE_AMBIGUOUS;
+        }
+        else {
+            my ( $h, $m ) = split /[:]/, $offset1;
+            $offset = $h * 60 + $m;
+        }
 
-    my ($name, $offset1, $offset2) = m< \A $EntryRx >x
-      or die qq/Could not parse zone entry at line: $./;
+        if ( exists $ZoneAbbrevs{$encoded} ) {
+            $flags |= DT_ZONE_AMBIGUOUS;
+            $offset = 0;
+        }
 
-    my $encoded = encode($name);
-    my $offset  = 0;
-    my $flags   = 0;
+        if ( exists $Universal{$name} ) {
+            $flags |= DT_ZONE_UTC;
+            $offset = 0;
+        }
 
-    if ($offset2) {
-        $flags  = DT_ZONE_AMBIGUOUS;
-    }
-    else {
-        my ($h, $m) = split /[:]/, $offset1;
-        $offset = $h * 60 + $m;
-    }
+        if ( exists $Military{$name} ) {
+            $flags |= DT_ZONE_MILITARY;
+            $offset = $Military{$name};
+        }
 
-    if (exists $Zone{$encoded}) {
-        $flags |= DT_ZONE_AMBIGUOUS;
-        $offset = 0;
-    }
+        if ( exists $Rfc{$name} ) {
+            $flags |= DT_ZONE_RFC;
+            $offset = $Rfc{$name};
+        }
 
-    if (exists $Universal{$name}) {
-        $flags |= DT_ZONE_UTC;
-        $offset = 0;
+        $ZoneAbbrevs{$encoded} = [ $flags, $offset, $name ];
     }
+}
 
-    if (exists $Military{$name}) {
-        $flags |= DT_ZONE_MILITARY;
-        $offset = $Military{$name};
-    }
+my %ZoneNames;
 
-    if (exists $Rfc{$name}) {
-        $flags |= DT_ZONE_RFC;
-        $offset = $Rfc{$name};
+sub read_zi_file($) {
+    my $filename = shift;
+    open( my $fh, '<:raw', $filename )
+      or die qq<Could not open '$filename' for reading: '$!'>;
+
+    my $zone;
+    my $format;
+    while (<$fh>) {
+        next if /^#/;
+        if (/^Zone\s+(\S+)\s+(\S+)\s+(\S+)\s+(\S+)/) {
+            $zone   = $1;
+            $format = $4;
+            $ZoneNames{$zone}++ if $format eq 'LMT';
+            next;
+        }
     }
 
-    $Zone{$encoded} = [$flags, $offset, $name];
 }
 
-printf "/* Automatically generated by tools/zone.pl */\n";
-foreach my $encoded (sort { $a <=> $b } keys %Zone) {
-    my ($flags, $offset, $name) = @{ $Zone{$encoded} };
-    
-    if ($flags != 0) {
-        my @names;
-
-        while ((my $flag = $flags & -$flags) != 0) {
-            push @names, $FlagName{$flag} || die sprintf '0x%.8X', $flag;
-            $flags &= ~$flag;
+sub gen_header() {
+    printf "/* Automatically generated by gen-zone-abbrevs.pl */\n";
+
+    # first we generate abbreviations list
+    foreach my $encoded ( sort { $a <=> $b } keys %ZoneAbbrevs ) {
+        my ( $flags, $offset, $name ) = @{ $ZoneAbbrevs{$encoded} };
+        my $nextid = next_id($name);
+
+        if ( $flags != 0 ) {
+            my @names;
+
+            while ( ( my $flag = $flags & -$flags ) != 0 ) {
+                push @names, $FlagName{$flag} || die sprintf '%4d', $flag;
+                $flags &= ~$flag;
+            }
+
+            $flags = join '|', @names;
         }
-        
-        $flags = join '|', @names;
+
+        printf "ZONE_ABBREV(%4d, %4d, \"%s\", %s)\n",
+          $nextid, $offset, $name, $flags;
     }
 
-    printf "DT_ZONE_ENTRY(0x%.8X, %4d, \"%s\", %s)\n",
-      $encoded, $offset, $name, $flags;
+    # second we enumerate all known from main.zi primary zone names
+    foreach my $zonename (sort keys %ZoneNames) {
+        my $nextid = next_id($zonename);
+        printf "ZONE_UNIQUE(%4d, \"%s\")\n", $nextid, $zonename;
+    }
 }
+
+read_abbrevs_file($zone_abbrevs_file);
+read_zi_file($tzdata_zi_file);
+gen_header();
```

Reply:
Yes, as you see I've significantly updated `zone.txt` to include all recent changes to the timezone abbreviations list (`c-dt` was not updated for more than decade).